### PR TITLE
Graphql

### PR DESCRIPTION
**Pull request by lorenpap**

### CODE REVIEW

File: src/db/products.ts [hunk 1]
```diff
@@ -22,5 +22,5 @@ const productSchema = new Schema({
         required: true
     }
 }, {versionKey: false});
-
+productSchema.set('toObject', {virtuals: true});
```
Comment (alon6699) [inline src/db/products.ts]: what does it do?

File: src/graphql/resolver.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import * as UserCart from '../socket/cart';
+import {getUpdatedProductsAmount} from '../socket/cart';
+import {getUpdatedProductAmount} from "../socket/socket-controllers";
+import {Product} from "../models/product";
+import {PubSub} from 'apollo-server-koa';
+
+const pubsub = new PubSub();
+
+export const updateProduct = (parent, {product}, context) => {
+    return pubsub.asyncIterator(['PRODUCT_UPDATED']);
```
Comment (alon6699) [inline src/graphql/resolver.ts]: put 'PRODUCT_UPDATED' into varaible

File: src/graphql/resolver.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import * as jwt from "jsonwebtoken";
+import * as UserCart from '../socket/cart';
+import {getUpdatedProductsAmount} from '../socket/cart';
+import {getUpdatedProductAmount} from "../socket/socket-controllers";
+import {Product} from "../models/product";
+import {PubSub} from 'apollo-server-koa';
+
+const pubsub = new PubSub();
+
+export const updateProduct = (parent, {product}, context) => {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: write in one line .. use arrow function advantage

File: src/graphql/resolver.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                });
+            }
+            return null;
+        },
+        getProduct: (parent, {_id}, context, info) => context.token ? queries.findProductByIdQuery(_id) : null
+    },
+    Mutation: {
+        addProduct: (parent, {product}, context, info) => {
+            if (context.token) {
+                return queries.addProductQuery(product).then(newProduct => {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: use aync await everywhere (remove then)

File: src/graphql/resolver.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                    pubsub.publish('PRODUCT_UPDATED', {
+                        productUpdated: deletedProduct.toObject()
+                    });
+                    return deletedProduct;
+                });
+            }
+            return null;
+        },
+        login: () => {
+            const token = jwt.sign({username: "ado"}, 'supersecret', {expiresIn: 1000});
```
Comment (alon6699) [inline src/graphql/resolver.ts]: the usename should change but you can leave it like this

File: src/graphql/resolver.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                    });
+                    return newProduct;
+                });
+            }
+            return null;
+        },
+        deleteProduct: async (parent, {_id}, context, info) => {
+            if (context.token) {
+                return await queries.deleteProductQuery(_id).then(deletedProduct => {
+                    pubsub.publish('PRODUCT_UPDATED', {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: move into a method ... create method that gets updatedProduct as arg and publishes it

File: src/graphql/resolver.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        },
+        updateProductAmount: (parent, {productId, selectedAmount}, context, info) => {
+            if (context.token) {
+                const userProduct: Partial<Product> = {
+                    _id: productId,
+                    amount: selectedAmount
+                };
+                UserCart.updateUsersProductsCache(userProduct, context.token);
+                pubsub.publish('PRODUCT_UPDATED', {
+                    productUpdated: getUpdatedProductAmount(context.token, productId)
```
Comment (alon6699) [inline src/graphql/resolver.ts]: do not calculate updatedProductAmount twice

File: src/graphql/resolver.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                });
+            }
+            return null;
+        },
+        login: () => {
+            const token = jwt.sign({username: "ado"}, 'supersecret', {expiresIn: 1000});
+            UserCart.createUserCart(token);
+            return token;
+        },
+        updateProductAmount: (parent, {productId, selectedAmount}, context, info) => {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: do not get args that you dont use .. for example remove info ... also you can rename parent to _ in order to make your code shorter:
updateProductAmount: (_, {productId, selectedAmount}, context) =>

File: src/graphql/resolver.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        },
+        checkout: (parent, args, context, info) => {
+            if (context.token) {
+                const usersProducts = UserCart.getUsersProducts();
+                return queries.checkoutProductQuery(usersProducts[context.token]).then(checkoutResponse => {
+                    UserCart.deleteUserCart(context.token);
+                    if (checkoutResponse instanceof Error) {
+                        return null;
+                    }
+                    checkoutResponse.forEach(product => {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: you dont need to publish products that already updated its unnecessary

File: src/graphql/type-defs.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import {gql} from "apollo-server";
+
+export const typeDefs = gql`
+type Product {
+    name: String!,
+    amount: Int!,
+    description: String!,
+    price: Int!,
+    _id: ID!
```
Comment (alon6699) [inline src/graphql/type-defs.ts]: change to id

File: src/graphql/type-defs.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+type Mutation {
+    addProduct(product: ProductInput): Product!,
+    deleteProduct(_id: ID): Product!,
+    login: ID!,
+    updateProductAmount(productId: ID, selectedAmount: Int): Product,
+    checkout: [Product!]
+},
+
+type Subscription {
+    productUpdated: PartialProduct!
```
Comment (alon6699) [inline src/graphql/type-defs.ts]: should return Product

File: src/graphql/type-defs.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+export const typeDefs = gql`
+type Product {
+    name: String!,
+    amount: Int!,
+    description: String!,
+    price: Int!,
+    _id: ID!
+}
+
+type PartialProduct {
```
Comment (alon6699) [inline src/graphql/type-defs.ts]: remove

File: src/graphql/type-defs.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+type Query {
+    getProducts: [Product!],
+    getProduct(_id: ID): Product!
+}
+
+type Mutation {
+    addProduct(product: ProductInput): Product!,
+    deleteProduct(_id: ID): Product!,
+    login: ID!,
+    updateProductAmount(productId: ID, selectedAmount: Int): Product,
```
Comment (alon6699) [inline src/graphql/type-defs.ts]: Product!

File: src/graphql/type-defs.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+type Query {
+    getProducts: [Product!],
+    getProduct(_id: ID): Product!
+}
+
+type Mutation {
+    addProduct(product: ProductInput): Product!,
+    deleteProduct(_id: ID): Product!,
```
Comment (alon6699) [inline src/graphql/type-defs.ts]: rename to id

File: src/graphql/type-defs.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+input UpdatedProductInput {
+    name: String,
+    amount: Int,
+    description: String,
+    price: Int,
+}
+
+type Query {
+    getProducts: [Product!],
+    getProduct(_id: ID): Product!
```
Comment (alon6699) [inline src/graphql/type-defs.ts]: id

File: src/graphql/resolver.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+            if (context.token) {
+                const userProduct: Partial<Product> = {
+                    _id: productId,
+                    amount: selectedAmount
+                };
+                UserCart.updateUsersProductsCache(userProduct, context.token);
+                pubsub.publish('PRODUCT_UPDATED', {
+                    productUpdated: getUpdatedProductAmount(context.token, productId)
+                });
+                return getUpdatedProductAmount(context.token, productId);
```
Comment (alon6699) [inline src/graphql/resolver.ts]: calculate products amount in products resolve ... each type including Product can have resolver that can manipulate its properties before sending the response.

File: src/index.ts [hunk 17]
```diff
@@ -7,41 +7,69 @@ import *  as nconf from 'nconf';
...
 
-server.listen(port, () => {
-        console.log(`✅  The server is running at http://localhost:${port}/`);
-        initDb();
-    }
-);
+app.use(errorHandler).use(log).use(cors(options)).use(respond()).use(bodyParser()).use(router.routes());
+
+// apolloServer.applyMiddleware({app});
+// const io = new Server(server);
```
Comment (alon6699) [inline src/index.ts]: do not put it in comment ... just dont pass server

File: src/index.ts [hunk 18]
```diff
@@ -7,41 +7,69 @@ import *  as nconf from 'nconf';
...
-});
 
-server.listen(port, () => {
-        console.log(`✅  The server is running at http://localhost:${port}/`);
-        initDb();
-    }
-);
+app.use(errorHandler).use(log).use(cors(options)).use(respond()).use(bodyParser()).use(router.routes());
+
+// apolloServer.applyMiddleware({app});
```
Comment (alon6699) [inline src/index.ts]: do you need this?

File: src/index.ts [hunk 19]
```diff
@@ -7,41 +7,69 @@ import *  as nconf from 'nconf';
...
+        onConnect: async (connectionParams, webSocket, context) => {
+            console.log(`Subscription client connected using Apollo server's built-in SubscriptionServer.`);
+        },
+        onDisconnect: async (webSocket, context) => {
+            console.log(`Subscription client disconnected.`);
+        }
+    },
+    context: context => {
+        const token = context.ctx ? context.ctx.request.header.authorization : '';
+        if (token) {
```
Comment (alon6699) [inline src/index.ts:36]: use method that you already wrote to validate token

File: src/graphql/resolver.ts [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+                    return getUpdatedProductsAmount(dbProducts);
+                });
+            }
+            return null;
+        },
+        getProduct: (parent, {_id}, context, info) => context.token ? queries.findProductByIdQuery(_id) : null
+    },
+    Mutation: {
+        addProduct: (parent, {product}, context, info) => {
+            if (context.token) {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: first throw exception if token not found. second think about other way to handle unauthorized mutations or queries

File: src/index.ts (hunk 19 above)
Comment (lorenpap) [reply src/index.ts:36]: the method i wrote gets ctx and next and cokies and here i use headers

File: src/graphql/resolver.ts [hunk 21]
```diff
@@ -0,0 +1,91 @@
+import * as queries from "../db/db-queries";
+import * as jwt from "jsonwebtoken";
+import * as UserCart from '../socket/cart';
+import {getUpdatedProductsAmount} from '../socket/cart';
+import {getUpdatedProductAmount} from "../socket/socket-controllers";
+import {Product} from "../models/product";
+import {AuthenticationError, PubSub} from 'apollo-server-koa';
+
+const pubsub = new PubSub();
+const trigger = 'PRODUCT_UPDATED';
```
Comment (alon6699) [inline src/graphql/resolver.ts]: rename.. what if u have 2 triggers?

File: src/graphql/resolver.ts [hunk 22]
```diff
@@ -0,0 +1,91 @@
+import * as queries from "../db/db-queries";
+import * as jwt from "jsonwebtoken";
+import * as UserCart from '../socket/cart';
+import {getUpdatedProductsAmount} from '../socket/cart';
```
Comment (alon6699) [inline src/graphql/resolver.ts]: its strange that your cart is inside socket folder.. move to utils folder

File: src/graphql/resolver.ts [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+};
+const publishUpdatedProduct = (updatedProduct) =>
+    pubsub.publish(trigger, {productUpdated: updatedProduct});
+
+export const resolvers = {
+    Product: {
+        id: (root) => root._id,
+        amount: async (root, args, context) => {
+            const updatedProductAmount =
+                await getUpdatedProductAmount(context.token, root.id);
```
Comment (alon6699) [inline src/graphql/resolver.ts]: why do you call method inside socket controllers? you just want to get products real amount calculated from the cart.

File: src/graphql/resolver.ts [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+            return updatedProductAmount.amount;
+        }
+    },
+    Query: {
+        getProducts: async (_, args, context) => {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: in our server all the methods declared in the top of the page and just called here in resolver declaration .. so you can read it clearly ... definde all your methods first and then use here like: 
const getProducts  = () => {....}
const getProduct= () => {....}
Query: {
    getProducts,
    getProduct
} 
(so getProducts and getProduct are methods defined above)

File: src/graphql/resolver.ts (hunk 20 above)
Comment (alon6699) [reply src/graphql/resolver.ts]: better solution would be: 
if anybody who is loged in to your server can access any query.mutation so the only mutation someone can access is login ... so you have two possible solution: 
1. move your login to be regular http nor graphql and then you throw exception in your graphql context if user isnt logged in.
2. validate token in graphql context only if the query is not login.

So you dont need to check if you have token each time here.
if you dont understant it you can call me

File: src/graphql/resolver.ts [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+export const resolvers = {
+    Product: {
+        id: (root) => root._id,
+        amount: (root) => {
+            const usersProducts = UserCart.getUsersProducts();
+            return usersProducts ? Object.keys(usersProducts).reduce((acc, curr) => {
+                return acc - (usersProducts[curr][root._id] || 0);
+            }, root.amount) : root.amount();
```
Comment (alon6699) [inline src/graphql/resolver.ts]: root.amount()? why ()?

File: src/graphql/resolver.ts [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    return checkoutResponse;
+};
+
+export const resolvers = {
+    Product: {
+        id: (root) => root._id,
+        amount: (root) => {
+            const usersProducts = UserCart.getUsersProducts();
+            return usersProducts ? Object.keys(usersProducts).reduce((acc, curr) => {
```
Comment (alon6699) [inline src/graphql/resolver.ts]: you have the same calculation in cart file ... reuse

**Pull request merged**